src/clr-angular/utils/id-generator/id-generator.spec.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import {Component, Inject, NgModule, Type} from "@angular/core";
+import {ComponentFixture, TestBed} from "@angular/core/testing";
+import {By} from "@angular/platform-browser";
+
+import {UNIQUE_ID, UNIQUE_ID_PROVIDER} from "./id-generator.service";
+
+@Component({
+    selector: "test-zero",
+    template: `
+        <div [id]="divId">My Div</div>`,
+    providers: [UNIQUE_ID_PROVIDER]
+})
+class IdZero {
+    constructor(@Inject(UNIQUE_ID) public divId: string) {}
+}
+
+@Component({
+    selector: "test-one",
+    template: `
+        <div [id]="divId">My Div</div>`,
+    providers: [UNIQUE_ID_PROVIDER]
+})
+class IdOne {
+    constructor(@Inject(UNIQUE_ID) public divId: string) {}
+}
+
+@Component({
+    selector: "test-two",
+    template: `
+        <div [id]="divId">My Div</div>`,
+    providers: [UNIQUE_ID_PROVIDER]
+})
+class IdTwo {
+    constructor(@Inject(UNIQUE_ID) public divId: string) {}
+}
+
+@NgModule({declarations: [IdZero, IdOne, IdTwo], exports: [IdZero, IdOne, IdTwo]})
+class IdTestingModule {}
+
+@Component({
+    template: `
+        <test-zero></test-zero>
+        <test-one></test-one>
+        <test-two></test-two>`
+})
+class UniqueIdTest {}
+
+interface TestContext<T extends UniqueIdTest|IdZero|IdOne|IdTwo> {
+    fixture: ComponentFixture<T>;
+    idZero: IdZero;
+    idOne: IdOne;
+    idTwo: IdTwo;
+}
+
+describe("Unique ID Generator Service", function() {
+    function setupTest<T>(testContext: TestContext<T>, testComponent: Type<T>) {
+        TestBed.configureTestingModule(
+            {imports: [IdTestingModule], providers: [UNIQUE_ID_PROVIDER], declarations: [testComponent]});
+        testContext.fixture = TestBed.createComponent(testComponent);
+        testContext.idZero = testContext.fixture.debugElement.query(By.directive(IdZero)).componentInstance;
+        testContext.idOne = testContext.fixture.debugElement.query(By.directive(IdOne)).componentInstance;
+        testContext.idTwo = testContext.fixture.debugElement.query(By.directive(IdTwo)).componentInstance;
+    }
+
+    it("generates uniq id's in the correct order", function() {
+        setupTest(this, UniqueIdTest);
+        expect(this.idZero !== this.idOne !== this.idTwo).toBe(true);
```

Comment:
This line doesn't work. This is interpreted in Javascript as 
```typescript
(this.idZero !== this.idOne) !== this.idTwo
```
So the first inequality returns a boolean, which is *always* `!==` from the last id.
You can easily try it out in your console, `1 !== 1 !== 1` is true.
What that means is this test always passes, even if the ids are the same. 😛 
On top of that, you're testing equality of the component instances, not their ids.

Anyway, based on my previous comments, what you want here is something like:
```typescript
const tests = [1, 2, 3];
for (const i=0; i<2; i++) {
    for (const j=i+1; j<3; j++) {
        expect(tests[i].divId).not.toEqual(tests[j].divId);
    }
}
```

Reply:
Updated.